engine/tests/phpunit/Elgg/ActionsServiceTest.php

```diff
@@ -34,67 +109,120 @@ public function testCanUnregisterActions($actions) {
-// 		}
+	public function testCanGenerateValidTokens() {
+
+		$timeout = $this->actions->getActionTokenTimeout();
+		for ($i = 1; $i <= 100; $i++) {
+			$timestamp = rand(time(), time() + $timeout);
+			$token = $this->actions->generateActionToken($timestamp);
+			$this->assertTrue($this->actions->validateActionToken(false, $token, $timestamp), "Test failed at pass $i");
+			$this->assertFalse($this->actions->validateActionToken(false, $token, $timestamp + 1), "Test failed at pass $i");
+			$this->assertFalse($this->actions->validateActionToken(false, $token, $timestamp - 1), "Test failed at pass $i");
+		}
+	}
+
+	public function testCanNotValidateExpiredToken() {
+		$timeout = $this->actions->getActionTokenTimeout();
+		$timestamp = time() - $timeout - 10;
+		$token = $this->actions->generateActionToken($timestamp);
+		$this->assertFalse($this->actions->validateActionToken(false, $token, $timestamp));
+	}
+
+	public function testCanNotValidateTokenAfterSessionExpiry() {
+		$timeout = $this->actions->getActionTokenTimeout();
+		$timestamp = time();
+		$token = $this->actions->generateActionToken($timestamp);
+		_elgg_services()->session->invalidate();
+		_elgg_services()->session->start();
+		$this->assertFalse($this->actions->validateActionToken(false, $token, $timestamp));
+	}
+
+	public function testCanGetValidActionTokenResponseData() {
+
+		$data = $this->actions->getActionTokenResponseData();
+
+		$this->assertArrayHasKey('__elgg_ts', $data);
+		$this->assertArrayHasKey('__elgg_token', $data);
+		$this->assertArrayHasKey('logged_in', $data);
+
+		$this->assertEquals($data['__elgg_token'], $this->actions->generateActionToken($data['__elgg_ts']));
+		$this->assertTrue($this->actions->validateActionToken(false, $data['__elgg_token'], $data['__elgg_ts']));
+	}
+
+	public function testActionGatekeeper() {
+
+		$result = $this->actions->gatekeeper('test');
+		$this->assertInstanceOf(RedirectResponse::class, $result);
+		$this->assertEquals($result, _elgg_services()->responseFactory->getSentResponse());
+
+		$this->passGatekeeper();
+		$this->assertTrue($this->actions->gatekeeper('test'));
+	}
+
+	public function testActionGatekeeperForLoginAction() {
+		// test action/login token validation
+		$this->markTestIncomplete();
+	}
+
+	public function testCanExecuteActionWithBufferOutput() {
+
+		// ajax forward hook
+		actions_init();
+
+		$this->request = $this->createRequest('action/buffer/success', 'POST', [], true, true);
+
+		$this->createService();
+
+		$this->assertTrue($this->actions->register('buffer/success', "$this->actionsDir/buffer/success.php", 'public'));
 		
-// 	}
-	
-	//TODO gatekeeper?
-}
+		$result = $this->actions->execute('buffer/success/', 'success');
+		$this->assertInstanceOf(ResponseBuilder::class, $result);
+
+		$this->assertEquals('success', $result->getForwardURL());
+		$this->assertEquals(ELGG_HTTP_OK, $result->getStatusCode());
+
+		_elgg_services()->router->route(_elgg_services()->request);
```

Comment:
@mrclay Any idea why I am getting Form is missing __token error on this route here? I can't figure out at which point it's lost from the Input.